<a href="https://colab.research.google.com/github/tristanreshef13/Data-Analysis-Portfolio/blob/main/pizza_sales_analsys_and_optimization_Using_Pandas_to_understand_the_data_tristan_reshef.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<div dir="rtl">

# **Assignment 2 - Using Pandas to Understand the Data**

### **תיאור המטלה**  
במטלה זו ניתחתי נתונים של **מכירות פיצרייה גדולה בארצות הברית** באמצעות **Pandas** בפייתון, במטרה לזהות **מגמות ולייעל תהליכים עסקיים**.

### **שלבי הניתוח**  
- **טעינת הנתונים** – קריאת קובץ הנתונים וסקירת המבנה שלו.  
- **ניקוי ועיבוד נתונים** – טיפול בערכים חסרים, הסרת כפילויות ושינוי סוגי נתונים.  
- **ניתוח סטטיסטי** – חקר מגמות מכירה לאורך זמן, זיהוי מוצרים פופולריים והשוואת ביצועים לפי סניפים.  
- **ויזואליזציה של נתונים** – יצירת גרפים ודאשבורדים להצגת תובנות מרכזיות.  
- **הסקת מסקנות** – זיהוי גורמים המשפיעים על המכירות והצעת דרכים לשיפור הרווחיות.

### **מטרות המטלה**  
✔ הבנה מעמיקה של **שימוש ב-Pandas** לניתוח נתונים עסקיים.  
✔ זיהוי **דפוסים במכירות** וניתוח הביצועים העסקיים של הפיצרייה.  
✔ שימוש בטכניקות שונות ב-**Data Analysis & Data Cleaning** כדי להפיק תובנות משמעותיות.

### **קבצים מצורפים**  
- **מאגר הנתונים (CSV)** – נתוני מכירות מהפיצרייה.  
- **Notebook זה** – כולל את כל שלבי הניתוח והקוד בפייתון.  

**הערה:** הניתוח מספק תובנות מעמיקות שעשויות לשפר את קבלת ההחלטות העסקיות של הפיצרייה.

</div>


# **שאלה 1**

**שאלה 1 - הכרות עם הדאטה** (25 נקודות)<br>
<br>
לאחר הורדת הנתונים, השתמשו בפאנדס ושמרו את הנתונים לתוך דאטאפריים<br>
הריצו פונקציות שונות כדי ללמוד על הנתונים בטבלה <br>
אל תשכחו להשאיר את הקוד וכל הפלט, כדי שנוכל לדעת על בסיס אילו נתונים אתם עונים
1. בתא טקסט תארו כמה תצפיות (שורות) וכמה משתנים (עמודות) יש לטבלה<br>
הסבירו על בסיס איזה קוד (דאגו כי יהיה פלט מודפס) מצאתם את התשובה
1. בתא טקסט תנו דוגמא למשתנה (עמודה) קטגוריאלי מכלל המשתנים הקיימים <br>
הסבירו מדוע הוא קטגוריאלי
1. בתא טקסט תנו דוגמא למשתנה (עמודה) כמותי מכלל המשתנים הקיימים<br>
1. בתא טקסט הסבירו:<br>
האם לדעתכם ישנן תצפיות שניתן להסיר מהנתונים?<br>אם כן, כתבו בתא טקסט שתי דוגמאות לתצפיות שהסיבה להסרתן מהנתונים **שונה**
1. אם תשובתכם לסעיף הקודם הייתה כן, ספקו בתא קוד את קטע הקוד באמצעותו הגעתם לתשובתכם

In [ ]:
from google.colab import drive
drive.mount('/content/drive')
import pandas as pd
df=pd.read_csv('/content/drive/MyDrive/pizza_sales.csv')
print(df.info())
print(df.describe())
print(df.head)

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48620 entries, 0 to 48619
Data columns (total 15 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Unnamed: 0         48620 non-null  int64  
 1   order_details_id   48620 non-null  int64  
 2   pizza_id           48620 non-null  object 
 3   quantity           43758 non-null  float64
 4   order_time         48620 non-null  object 
 5   unit_price         44077 non-null  float64
 6   total_price        48620 non-null  float64
 7   pizza_size         43758 non-null  object 
 8   pizza_type         48620 non-null  object 
 9   pizza_ingredients  48620 non-null  object 
 10  pizza_name         43758 non-null  object 
 11  is_it_pizza        48620 non-null  object 
 12  day_of_week        48620 non-null  object 
 13  time_of_day        48620 non-null  ob

1. We have **48,620 rows** and **15 columns**, which can be observed using the `info()` function.

2. The `pizza_size` column is categorical because it represents distinct categories of pizza sizes, such as "Small", "Medium", and "Large". These are non-numeric values that indicate different types of pizzas, and mathematical operations like averages cannot be performed on them. Therefore, this column contains discrete values, meaning each value belongs to a specific category or group rather than being part of a continuous scale.

3. The `unit_price` column is quantitative because it contains numerical values representing the price of a single unit of pizza. The values in this column are continuous, allowing calculations such as mean, median, or sum, which makes it a quantitative column.


4. Here are two categories of problematic observations that I suggest could be removed:

**1. Missing or Zero Values:**  
- **Missing Quantity (`quantity`)**: When the number of pizzas ordered is missing (shown as NaN in the "quantity" column), it’s difficult to accurately analyze sales and revenue. In such cases, I suggest removing these rows to improve the reliability of the analysis.  
  *Example*: A row where the quantity is missing.  
- **Zero Unit Price (`unit_price`)**: A unit price of zero may indicate either an error or a special promotion. If it’s due to an error, removing it would prevent biased results. If it’s a promotion, removal should be considered if it doesn’t contribute meaningfully to the overall analysis.  
  *Example*: A row where `unit_price` equals zero.

**2. Unrealistic or Extreme Prices:**  
- **Unrealistic `unit_price` or `total_price`**: Extremely high or unrealistic prices could signal data errors. These rows should be carefully reviewed and, if necessary, removed to prevent them from distorting the analysis.  
  *Example*: A row with a `total_price` of 8300, which seems unusually high compared to other rows and requires further investigation.

In [ ]:
#שאלה 1 סעיף 5
missing_values = df.isna().sum()
print("Number of missing values in each column:")
print(missing_values)
print(df.describe())

Number of missing values in each column:
Unnamed: 0              0
order_details_id        0
pizza_id                0
quantity             4862
order_time              0
unit_price           4543
total_price             0
pizza_size           4862
pizza_type              0
pizza_ingredients       0
pizza_name           4862
is_it_pizza             0
day_of_week             0
time_of_day             0
payment_method          0
dtype: int64
         Unnamed: 0  order_details_id      quantity    unit_price  \
count  48620.000000      48620.000000  43758.000000  44077.000000   
mean   24309.500000      24310.500000      1.019585     23.173808   
std    14035.529381      14035.529381      0.142313     50.762232   
min        0.000000          1.000000      1.000000      0.000000   
25%    12154.750000      12155.750000      1.000000     12.500000   
50%    24309.500000      24310.500000      1.000000     16.500000   
75%    36464.250000      36465.250000      1.000000     20.250000   
max 

# **שאלה 2**

**שאלה 2 - השוואה בין סוגי פיצות** (24 נקודות)<br>
1. הדפיסו כמה פיצות יש מכל סוג <br>
Pizza_type<br>
2. בעלי הפיצרייה החליטו להשוות רק בין סוגי הפיצות שהן לא מוגדרות כטבעוניות<br>
'Classic', 'Supreme', 'Chicken'<br>
 צרו דאטהפריים חדש רק עם סוגים אלו<br>

1. הסירו את כל הטורים שלא מוסיפים מידע להשוואה בין הקבוצות, שימרו את הדאטה פריים המעודכן<br>
בתא טקסט, הסבירו עבור כל טור למה הורדתם אותו<br>

1. בתא קוד, חקרו את ההבדלים בין סוגי הפיצות בדאטה החדש
באמצעות שתי שיטות סטטיסטיות שתבחרו<br>
הסבירו את התוצאות של הסטטיסטיקות שאתם רואים וציינו מדוע בחרתם לחקור באמצעות שיטות אלו <br>שימו לב כי ניתן לפתור סעיף זה ביותר מדרך אחת<br>(שמירת דאטאפריים נפרד לכל סוג פיצה או שימוש בפעולת<br>groupby)<br>


In [ ]:
#שאלה 2 סעיף 1
print(df['pizza_type'].value_counts())


pizza_type
Classic    14579
Supreme    11777
Veggie     11449
Chicken    10815
Name: count, dtype: int64


In [ ]:
#שאלה 2 סעיף 2
df_no_vegen=df[df['pizza_type'] != 'Veggie']


In [ ]:
#שאלה 2 סעיף 3
# הסרת טורים שלא מוסיפים מידע להשוואה בין סוגי הפיצות
df_clean= df_no_vegen.drop(columns=['Unnamed: 0', 'order_details_id', 'order_time', 'pizza_name', 'is_it_pizza'])




**Unnamed: 0** – This is a redundant column as it only counts the rows and does not contain any useful information.  
**order_details_id** – A unique identifier for each order. It has no comparative value between different types of pizzas.  
**order_time** – This column indicates the order time, but for the purpose of comparing pizza types (and not order times), it is irrelevant.  
**pizza_name** – The pizza names might be unique to each pizza type, and therefore they do not add any additional information beyond the pizza_type column and the ingredians column.  
**is_it_pizza** – All rows contain the value "Yes," so this column has no value when comparing different pizza types.

In [ ]:
#שאלה 2 סעיף 4
# חישוב ממוצע, סטיית תקן ומספר התצפיות לפי סוג פיצה וכמות
price_stats = df_clean.groupby(['pizza_type', 'quantity']).agg(
    avg_total_price=('total_price', 'mean'),
    std_total_price=('total_price', 'std'),
    count=('total_price', 'count')
).reset_index()

# הדפסת התוצאות
print("Statistics (mean, std, count) by pizza type and quantity:")
print(price_stats)



Statistics (mean, std, count) by pizza type and quantity:
   pizza_type  quantity  avg_total_price  std_total_price  count
0     Chicken       1.0       190.104180      1185.673477   9558
1     Chicken       2.0       156.835784       997.338753    204
2     Chicken       3.0        56.250000         6.928203      4
3     Chicken       4.0        83.000000              NaN      1
4     Classic       1.0       175.963976      1147.050195  12843
5     Classic       2.0       292.881526      1461.820949    249
6     Classic       3.0        41.571429         9.825259      7
7     Classic       4.0        48.000000              NaN      1
8     Supreme       1.0       169.293997      1114.725693  10462
9     Supreme       2.0       168.626486      1046.842503    185
10    Supreme       3.0        39.450000        23.732494      5


#התגלו לנו סטיות תקן מאוד קיצוניות עבור סקטור האוכל,נצטרך לבחון את הדאטה יותר לעומק ולנקות אותו

In [ ]:
df2=df_clean[df_clean['total_price'] == df_clean['quantity'] * df_clean['unit_price']]
df2= df2[(df2['total_price'] >0)]
df2.describe()
#אותו תהליך ניקוי שעשיתי בשאלה הבאה

quantity    unit_price   total_price
count  26130.000000  26130.000000  26130.000000
mean       1.020819     16.462732     16.806529
std        0.147786      3.801684      4.618106
min        1.000000      9.750000      9.750000
25%        1.000000     12.500000     12.500000
50%        1.000000     16.500000     16.500000
75%        1.000000     20.750000     20.750000
max        4.000000     35.950000     83.000000

In [ ]:
# חישוב ממוצע, סטיית תקן, חציון ומספר תצפיות לפי סוג פיצה וכמות
price_stats =df2.groupby(['pizza_type', 'quantity']).agg(
    avg_unit_price=('unit_price', 'mean'),
    std_unit_price=('unit_price', 'std'),
    median_unit_price=('unit_price', 'median'),
    count=('unit_price', 'count')
).reset_index()

# חישוב סך כל התצפיות
total_count = price_stats['count'].sum()

# הוספת עמודת אחוז מכלל התצפיות
price_stats['percent_of_total'] = (price_stats['count'] / total_count) * 100

print("Statistics (mean, std, median, count, percent) by pizza type and quantity:")
print(price_stats)


Statistics (mean, std, median, count, percent) by pizza type and quantity:
   pizza_type  quantity  avg_unit_price  std_unit_price  median_unit_price  \
0     Chicken       1.0       17.668251        3.078187              16.75   
1     Chicken       2.0       18.506098        2.702242              20.75   
2     Chicken       3.0       18.083333        2.309401              16.75   
3     Chicken       4.0       20.750000             NaN              20.75   
4     Classic       1.0       14.812852        3.969128              13.25   
5     Classic       2.0       14.020202        3.557661              12.00   
6     Classic       3.0       12.750000        1.604681              12.00   
7     Classic       4.0       12.000000             NaN              12.00   
8     Supreme       1.0       17.358966        3.461536              16.50   
9     Supreme       2.0       17.874333        3.353330              16.50   
10    Supreme       3.0       16.437500        3.375000            

הערה:לא הצלחתי להבין אם מדובר  בשתי שיטות סטטסטיות בקוד או שני אומדנים סטטסטיים,אז רשמתי יותר ליתר ביטחון ואפשר לקרוא רק את השניים הראשונים במידת הצורך

From the obtained data, the following insights can be drawn regarding the differences between pizza types based on the presented statistics:

1. **Average Unit Price (`avg_unit_price`)**  
   Chicken pizzas are the most expensive on average (17.67–20.75) compared to Classic pizzas (12.0–14.81).  
   Supreme pizzas fall within a price range of 16.44 to 17.87, positioning them between Classic and Chicken pizzas.

2. **Standard Deviation of Unit Price (`std_unit_price`)**  
   The standard deviation is higher for Classic pizzas (1.60–3.96), indicating greater price variability. This could be due to more customization options or a higher frequency of extra toppings.  
   Chicken and Supreme pizzas have a lower standard deviation (2.30–3.46), suggesting more consistent pricing across different orders.

3. **Median Unit Price (`median_unit_price`)**  
   The median price is close to the average in most cases, showing that prices are not heavily skewed.  
   For Classic pizzas, in some cases (e.g., orders of 1 or 2 pizzas), the median is lower than the average, indicating the presence of lower-priced orders.

4. **Percentage of Total Orders (`percent_of_total`)**  
   Classic pizzas are the most popular, accounting for 38.05% of total orders, followed by Supreme pizzas (31.13%) and Chicken pizzas (28.79%).  
   Orders with more than one pizza are rare, representing less than 2% of total orders.

**שאלה 3 - קשר בין מחיר הפיצה למשתנים אחרים**(24 נקודות)<br>
בעלי הפיצרייה החליטו לחקור את הקשר בין מחיר הפיצה לבין משתנים אחרים
1. הם חושבים שחלק מהתצפיות (שורות) לא רלוונטיות לניתוח <br>
השתמשו בפקודה <br>
describe <br>
על מנת לבדוק אלו תצפיות הייתם מסירים מניתוח הנתונים
<br>
כתבו בתא טקסט אלו תצפיות הייתם מסירים
<br>
 לאחר מכן, הגדירו קריטריון (כלל)  לפיו תסננו את התצפיות מהטבלה<br>
וצרו דאטהפריים חדש המכיל רק תצפיות מתאימות והמשיכו לעבוד איתו<br>
2.בנוסף, בעלי הפיצריה החליטו לבחון הבדלים בין פיצות בעלות מחיר גבוה יחסית לכל השאר.<br>
ראשית הם רצו להגדיר מהו מחיר גבוה<br>
בתא טקסט הסבירו לפי איזה קריטריון הייתם מגדירים פיצה עם מחיר גבוה<Br>
1.(יקר או זול) צרו משתנה בוליאני חדש המציין האם הפיצה בעלת מחיר גבוה או לא<br>
לאחר מכן, הגדירו עמודה חדשה בדאטאפריים המציינת האם הפיצה בעלת מחיר יקר או זול
<br>
1. השוו בין פיצות עם מחיר גבוה ונמוך<br>
בתא טקסט נפרד תארו 2 הבדלים<br>






In [ ]:
#שאלה 3 סעיף 1 חקירה
df.describe()
# Sort the DataFrame by 'unit_price' in descending order
df_sorted_by_unit_price_desc = df.sort_values(by='unit_price', ascending=False)

# Display the first few rows of the sorted DataFrame
print("Data sorted by unit price in descending order:")
print(df_sorted_by_unit_price_desc.head())

Data sorted by unit price in descending order:
       Unnamed: 0  order_details_id       pizza_id  quantity order_time  \
39883       39883             39884  classic_dlx_m       1.0   19:24:32   
26581       26581             26582  ital_veggie_s       1.0   13:35:26   
13545       13545             13546     thai_ckn_m       NaN   17:32:27   
8890         8890              8891  ckn_alfredo_l       1.0   16:09:05   
44023       44023             44024   veggie_veg_l       1.0   17:33:44   

       unit_price  total_price pizza_size pizza_type  \
39883       359.5        16.00          M    Classic   
26581       359.5        12.75          S     Veggie   
13545       359.5        16.75          M    Chicken   
8890        359.5        20.75          L    Chicken   
44023       359.5        20.25          L     Veggie   

                                       pizza_ingredients  \
39883  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
26581  Eggplant, Artichokes, Tomatoes, Zucchi

תוכנית לניקוי הנתונים:כפי שהוצע בשאלה 1

הסרת שורות עם ערכים חסרים בעמודות חיוניות:
יש להסיר את כל השורות שבהן קיימים ערכים חסרים בעמודות קריטיות לניתוח, כגון מספר הפיצות, המחיר הכולל והמחיר ליחידה.

הסרת שורות עם ערכים בלתי סבירים:
יש למחוק שורות שבהן המחיר הכולל או המחיר ליחידה שווה ל-0, מכיוון שערכים אלו אינם הגיוניים וסביר להניח שמדובר בטעויות או נתונים חסרים.

סקירת שורות עם מחירים כוללים גבוהים במיוחד:
יש לבדוק שורות שבהן המחיר הכולל חריג בגובהו, על מנת להחליט אם מדובר בהזמנות גדולות אמיתיות או בשגיאות הקלדה.

In [ ]:
#שאלה 3 סעיף 1 ניקוי דאטה
df_without_mistakes_question3 = df[df['total_price'] == df['quantity'] * df['unit_price']]
df_without_mistakes_question3 = df_without_mistakes_question3[(df_without_mistakes_question3['total_price'] >0)]
df_without_mistakes_question3.describe()

Unnamed: 0  order_details_id      quantity    unit_price  \
count  34091.000000      34091.000000  34091.000000  34091.000000   
mean   24221.592414      24222.592414      1.019829     16.495248   
std    14031.041239      14031.041239      0.143562      3.623201   
min        1.000000          2.000000      1.000000      9.750000   
25%    12084.500000      12085.500000      1.000000     12.750000   
50%    24229.000000      24230.000000      1.000000     16.500000   
75%    36331.500000      36332.500000      1.000000     20.250000   
max    48619.000000      48620.000000      4.000000     35.950000   

        total_price  
count  34091.000000  
mean      16.826391  
std        4.445846  
min        9.750000  
25%       12.750000  
50%       16.500000  
75%       20.500000  
max       83.000000

In my opinion, an expensive pizza is one that costs more than 50% of the pizzas of the same size at the store. Based on the results from our describe() function, we can see that for a small pizza, an expensive pizza costs 12 or more, for a medium pizza, it costs 16.25 or more, and for a large pizza, it costs 20.5 or more per unit.

In [ ]:
#שאלה 3 סעיף 2
# Filter the DataFrame to include only small, medium, and large pizzas
small_pizzas = df_without_mistakes_question3[df_without_mistakes_question3['pizza_size'] == 'S']
print("Basic statistics for small pizzas (S):")
print(small_pizzas.describe())

medium_pizzas = df_without_mistakes_question3[df_without_mistakes_question3['pizza_size'] == 'M']
print("\nBasic statistics for medium pizzas (M):")
print(medium_pizzas.describe())

large_pizzas = df_without_mistakes_question3[df_without_mistakes_question3['pizza_size'] == 'L']
print("\nBasic statistics for large pizzas (L):")
print(large_pizzas.describe())


Basic statistics for small pizzas (S):
         Unnamed: 0  order_details_id     quantity   unit_price  total_price
count   9429.000000       9429.000000  9429.000000  9429.000000  9429.000000
mean   24358.163856      24359.163856     1.018136    12.371678    12.594703
std    14079.942431      14079.942431     0.138901     2.259993     2.875103
min       10.000000         11.000000     1.000000     9.750000     9.750000
25%    12010.000000      12011.000000     1.000000    12.000000    12.000000
50%    24589.000000      24590.000000     1.000000    12.000000    12.000000
75%    36424.000000      36425.000000     1.000000    12.500000    12.500000
max    48619.000000      48620.000000     4.000000    23.650000    48.000000

Basic statistics for medium pizzas (M):
         Unnamed: 0  order_details_id      quantity    unit_price  \
count  10242.000000      10242.000000  10242.000000  10242.000000   
mean   24152.611404      24153.611404      1.017087     15.960408   
std    13955.991615 

In [ ]:
#שאלה 3 סעיף 3
expensive_price_limit_s= 12
expensive_price_limit_m=16.25
expensive_price_limit_l=20.5
is_expensive = (  #משתנה בוליאני
    (df_without_mistakes_question3['unit_price'] >= expensive_price_limit_s) & (df_without_mistakes_question3['pizza_size'] == 'S') |
    (df_without_mistakes_question3['unit_price'] >= expensive_price_limit_m) & (df_without_mistakes_question3['pizza_size'] == 'M') |
    (df_without_mistakes_question3['unit_price'] >= expensive_price_limit_l) & (df_without_mistakes_question3['pizza_size'] == 'L')
)
df_without_mistakes_question3['is_expensive'] = is_expensive #הוספת עמודה בדאטה
df_without_mistakes_question3.head()


Unnamed: 0  order_details_id       pizza_id  quantity order_time  \
1           1                 2  classic_dlx_m       1.0   11:57:40   
5           5                 6     thai_ckn_l       1.0   11:57:40   
7           7                 8   prsc_argla_l       1.0   12:12:28   
8           8                 9    ital_supr_m       1.0   12:16:31   
9           9                10    ital_supr_m       1.0   12:21:30   

   unit_price  total_price pizza_size pizza_type  \
1       16.00        16.00          M    Classic   
5       20.75        20.75          L    Chicken   
7       20.75        20.75          L    Supreme   
8       16.50        16.50          M    Supreme   
9       16.50        16.50          M    Supreme   

                                   pizza_ingredients  \
1  Pepperoni, Mushrooms, Red Onions, Red Peppers,...   
5  Chicken, Pineapple, Tomatoes, Red Peppers, Tha...   
7  Prosciutto di San Daniele, Arugula, Mozzarella...   
8  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   
9  Calabrese Salami, Capocollo, Tomatoes, Red Oni...   

                         pizza_name is_it_pizza day_of_week time_of_day  \
1          The Classic Deluxe Pizza         Yes    Thursday     morning   
5            The Thai Chicken Pizza         Yes    Thursday     morning   
7  The Prosciutto and Arugula Pizza         Yes      Friday        noon   
8         The Italian Supreme Pizza         Yes    Thursday        noon   
9         The Italian Supreme Pizza         Yes    Thursday        noon   

  payment_method  is_expensive  its_weekend  
1    Credit Card         False        False  
5           Cash          True        False  
7           Cash          True        False  
8    Credit Card          True        False  
9           Cash          True        False

In [ ]:
#שאלה 3 סעיף 4
df_expensive_pizzas_only = df_without_mistakes_question3[df_without_mistakes_question3['is_expensive'] == True].copy()
df_cheap_pizzas_only = df_without_mistakes_question3[df_without_mistakes_question3['is_expensive'] == False].copy()
# השוואנו בין אמצעי התשלום ומצאנו כי ההבדל זניח



In [ ]:
#בודקים באיזה סוג פיצה יש יותר תוספות
#נחלק אתרשימת הרכיבים לרשימה ונספור את כמות האייטמים ברשימה
df_expensive_pizzas_only['num_ingredients'] = df_expensive_pizzas_only['pizza_ingredients'].str.split(',').apply(len)
df_cheap_pizzas_only['num_ingredients'] = df_cheap_pizzas_only['pizza_ingredients'].str.split(',').apply(len)
#נחשב :
average_ingredients1 = df_expensive_pizzas_only['num_ingredients'].mean()
print("\nAverage number of ingredients in expensive pizzas:", average_ingredients1)

average_ingredients2 = df_cheap_pizzas_only['num_ingredients'].mean()
print("\nAverage number of ingredients in cheap pizzas:", average_ingredients2)

print(df_expensive_pizzas_only.describe())
print(df_cheap_pizzas_only.describe())




Average number of ingredients in expensive pizzas: 5.875879228671228

Average number of ingredients in cheap pizzas: 4.933747725894482
        Unnamed: 0  order_details_id      quantity    unit_price  \
count  20899.00000       20899.00000  20899.000000  20899.000000   
mean   24226.87119       24227.87119      1.019666     16.569838   
std    14034.06831       14034.06831      0.143930      3.660987   
min        5.00000           6.00000      1.000000     12.000000   
25%    12057.00000       12058.00000      1.000000     12.500000   
50%    24215.00000       24216.00000      1.000000     16.750000   
75%    36391.50000       36392.50000      1.000000     20.750000   
max    48619.00000       48620.00000      4.000000     23.650000   

        total_price  num_ingredients  
count  20899.000000     20899.000000  
mean      16.900699         5.875879  
std        4.497362         1.183031  
min       12.000000         3.000000  
25%       12.500000         5.000000  
50%       16.7500

1. **We can see that expensive pizzas, on average, have more ingredients.**

2. **Based on this data (`describe()`), customers buy more units of expensive pizzas (20,899) than units of cheap pizzas (13,192).**

**שאלה 4 - הבדלים בקניית פיצות כתלות במועד הקנייה**(12 נקודות)
<br>
לאחר ההשוואה בין סוגי הפיצות,
בעלי הפיצרייה רצו לבדוק האם יש הבדל בין קניית פיצות כתלות במועד הקנייה
<br>
ראשית, בעלי הפיצרייה רצו להשוות בין פיצות הנקנות בסוף שבוע לעומת פיצות הנקנות בימי חול
1. על מנת לעזור לבעלי הפיצרייה לבדוק זאת, הוסיפו לטבלה עמודה המספקת מידע בנוגע למועד קניית הפיצה
<br>
(הוסיפו עמודה המתארת משתנה בוליאני - האם הפיצה נקנתה בסוף שבוע או לא)
<br>
2. לאחר מכן, בעלי הפיצרייה רצו לבחון הבדלים בין פיצות הנקנות בין סוף השבוע לאלו הנקנות בימי חול
<br>
האם ניתן למצוא הבדל משמעותי בין פיצות הנקנות בסוף השבוע לאלו הנקנות בימי חול? אם כן, כתבו בתא טקסט מהו ההבדל המשמעותי ובתא קוד הראו כיצד הגעתם לכך

# **שאלה 4**

In [ ]:
df4=df_without_mistakes_question3
df4['its_weekend']=df4['day_of_week'].isin(['Saturday', 'Sunday'])
df_weekend=df4[df4['its_weekend']==True]
df_weekday=df4[df4['its_weekend']==False]
print(df_weekend.describe())
print(df_weekday.describe())
print("quantity of pizza units devided by two (satarday,sunday)=",13081/2)
print("quantity of pizza units devided by 5 days of the week=",21010/5)
print("we can see that people let themselfs order more pizza units on the weekend days then regular days")


         Unnamed: 0  order_details_id      quantity    unit_price  \
count  13081.000000      13081.000000  13081.000000  13081.000000   
mean   24015.967281      24016.967281      1.020182     17.473083   
std    13963.056619      13963.056619      0.144383      3.619242   
min       20.000000         21.000000      1.000000      9.750000   
25%    11880.000000      11881.000000      1.000000     15.250000   
50%    24086.000000      24087.000000      1.000000     17.950000   
75%    36161.000000      36162.000000      1.000000     20.750000   
max    48618.000000      48619.000000      4.000000     35.950000   

        total_price  
count  13081.000000  
mean      17.833002  
std        4.565731  
min        9.750000  
25%       15.250000  
50%       17.950000  
75%       20.750000  
max       83.000000  
         Unnamed: 0  order_details_id      quantity    unit_price  \
count  21010.000000      21010.000000  21010.000000  21010.000000   
mean   24349.616326      24350.616326     

1. **We can see that, on average, expensive pizzas have more ingredients.**

2. **Based on this data (`describe()`), customers purchase more units of expensive pizzas (20,899) than units of cheaper pizzas (13,192).**

**שאלה 5 - שאלה פתוחה**(15 נקודות)
<br>
!כעת, בעלי הפיצרייה מבקשים את עזרתכם בניתוח נתונים שיעזור להם לקבל תובנות עסקיות
<br>
1. חשבו על ניתוח נוסף לנתונים שטרם נעשה ויוכל לסייע לבעלי הפיצרייה בקבלת תובנות עסקיות
<br>
בתיבת טקסט כתבו איזה ניתוח הייתם מבצעים
2. כעת, בתיבת קוד, ממשו את ניתוח הנתונים שהצעתם וכתבו את ההמלצה העסקית בעקבותיו

### **שאלה 5**

I would like to better understand the peak hours and peak days to optimize labor costs using the data. For example, on Mondays, people may prefer homemade meals and order less pizza, allowing me to schedule fewer workers for the Monday shift. On the other hand, Fridays might have higher pizza orders, especially in the morning, requiring more staff. Only data can reveal these trends, so we will analyze it to uncover these patterns.

We start by checking how many pizzas the company produces each day to get a clearer picture. For this, we use `zip` – a function that combines two (or more) lists into pairs. I learned about this function from ChatGPT. Additionally, I used `.copy()` to create a shallow copy of the data. Since I wasn’t sure where the research would lead and was experimenting with unfamiliar functions, I chose to avoid risks by ensuring the original data remained unchanged, allowing for safe testing.

In [ ]:
df5 = df_without_mistakes_question3
#נשתמש בקופי-פונקציה שיוצרת דאטה בייס חדש שלא משפיע על
#הדאטה הקודם כדי למנוע טעויות שישנו לנו דברים בדאטה המקורי
df_monday = df5[df5['day_of_week'] == 'Monday'].copy()
df_tuesday = df5[df5['day_of_week'] == 'Tuesday'].copy()
df_wednesday = df5[df5['day_of_week'] == 'Wednesday'].copy()
df_thursday = df5[df5['day_of_week'] == 'Thursday'].copy()
df_friday = df5[df5['day_of_week'] == 'Friday'].copy()
df_saturday = df5[df5['day_of_week'] == 'Saturday'].copy()
df_sunday = df5[df5['day_of_week'] == 'Sunday'].copy()

def print_time_of_day_and_quantity():
    # List of DataFrames for each day of the week
    days_data = [df_monday, df_tuesday, df_wednesday, df_thursday, df_friday, df_saturday, df_sunday]
    days_names = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']


    for day_df, day_name in zip(days_data, days_names):
        print(f"Total quantity of pizzas for each time of day on {day_name}:")

        # Multiply 'quantity' by 1 to get the total number of pizzas ordered for each time of day
        total_pizzas = day_df.groupby('time_of_day')['quantity'].sum()

        # Calculate the total pizzas (morning + afternoon + evening)
        total_pizzas_all_times = total_pizzas.sum()  # Sum of morning, afternoon, and evening

        print(total_pizzas)
        print(f"Total pizzas for {day_name} (morning + afternoon + evening): {total_pizzas_all_times}")
        print("\n" + "-"*40)

print_time_of_day_and_quantity()


Total quantity of pizzas for each time of day on Monday:
time_of_day
evening     998.0
morning     229.0
noon       2038.0
Name: quantity, dtype: float64
Total pizzas for Monday (morning + afternoon + evening): 3265.0

----------------------------------------
Total quantity of pizzas for each time of day on Tuesday:
time_of_day
evening    1060.0
morning     202.0
noon       2201.0
Name: quantity, dtype: float64
Total pizzas for Tuesday (morning + afternoon + evening): 3463.0

----------------------------------------
Total quantity of pizzas for each time of day on Wednesday:
time_of_day
evening    1066.0
morning     287.0
noon       2210.0
Name: quantity, dtype: float64
Total pizzas for Wednesday (morning + afternoon + evening): 3563.0

----------------------------------------
Total quantity of pizzas for each time of day on Thursday:
time_of_day
evening    1071.0
morning     238.0
noon       2423.0
Name: quantity, dtype: float64
Total pizzas for Thursday (morning + afternoon + evening

Now, we will create a similar code that calculates the percentage of our budget to be allocated to salaries for each shift on each day. This code is based on the previous one, with the necessary adjustments to provide the required information.

In [ ]:
!pip install colorama

In [ ]:
from colorama import Fore, Style
total_pizzas_week = df5['quantity'].sum()
def salary_budget_distribution():
    # List of DataFrames for each day of the week
    days_data = [df_monday, df_tuesday, df_wednesday, df_thursday, df_friday, df_saturday, df_sunday]
    days_names = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

    # Define rainbow colors
    rainbow_colors = [Fore.RED, Fore.YELLOW, Fore.GREEN, Fore.CYAN, Fore.BLUE, Fore.MAGENTA]

    # Iterate through each DataFrame
    for day_df, day_name in zip(days_data, days_names):
        print(f"{rainbow_colors[0]}Total quantity of pizzas for each time of day on {day_name}:{Style.RESET_ALL}")

        # Calculate the total number of pizzas ordered for each time of day (morning, afternoon, evening)
        total_pizzas = day_df.groupby('time_of_day')['quantity'].sum()

        # Calculate the total pizzas (morning + afternoon + evening)
        total_pizzas_daily = total_pizzas.sum()

        # Calculate the percentage of each time of day relative to the total pizzas for the week
        percentage_pizzas = total_pizzas_daily / total_pizzas_week * 100

        # Calculate the percentage of total pizzas for the day relative to the weekly total
        percentage_day = total_pizzas_daily / total_pizzas_week * 100

        # Print the total pizzas for each time of day and the percentage of the weekly total
        print(f"{rainbow_colors[1]}{total_pizzas}{Style.RESET_ALL}")
        print(f"{rainbow_colors[2]}Total pizzas for {day_name} (morning + afternoon + evening): {total_pizzas_daily}{Style.RESET_ALL}")
        print(f"{rainbow_colors[3]}The daily pizza units divided by the weekly pizza amount would give us the percentage of budget we want to allocate to that day/amount of time.{Style.RESET_ALL}")
        print(f"{rainbow_colors[4]}Percentage of total budget that should go to salaries on {day_name}:{Style.RESET_ALL}")
        print(f"{rainbow_colors[5]}{percentage_pizzas}{Style.RESET_ALL}")
        print(f"{rainbow_colors[0]}Percentage of total pizzas for the week on {day_name}: {percentage_day:.2f}%{Style.RESET_ALL}")
        print("\n" + "-" * 40)

salary_budget_distribution()


Total quantity of pizzas for each time of day on Monday:
time_of_day
evening     998.0
morning     229.0
noon       2038.0
Name: quantity, dtype: float64
Total pizzas for Monday (morning + afternoon + evening): 3265.0
The daily pizza units divided by the weekly pizza amount would give us the percentage of budget we want to allocate to that day/amount of time.
Percentage of total budget that should go to salaries on Monday:
9.391089251301521
Percentage of total pizzas for the week on Monday: 9.39%

----------------------------------------
Total quantity of pizzas for each time of day on Tuesday:
time_of_day
evening    1060.0
morning     202.0
noon       2201.0
Name: quantity, dtype: float64
Total pizzas for Tuesday (morning + afternoon + evening): 3463.0
The daily pizza units divided by the weekly pizza amount would give us the percentage of budget we want to allocate to that day/amount of time.
Percentage of total budget that should go to salaries on Tuesday:
9.960594816924093
Percenta

**Insights**

**Days with High Order Volume:**  
Friday (21.28%), Saturday (20.25%), and Sunday (18.14%) together account for nearly 60% of the weekly orders – a higher budget for staff allocation should be prioritized on these days.

**Days with Low Order Volume:**  
Monday (9.39%) and Tuesday (9.96%) have the lowest order volume – staff budget can be reduced on these days.

**Key Shifts:**  
Noon is the busiest shift every day, especially on Friday, Saturday, and Sunday.  
Evening is also significant, mainly on Friday and Saturday.  
In contrast, morning shifts account for a very low percentage of the total weekly orders.


**Recommendation:**  
Prioritize more staff during noon and evening shifts on Friday, Saturday, and Sunday.  
Reduce staff during morning shifts and overall on Monday and Tuesday due to lower demand. Use my percentage module to divide your workers and update the data regularly to ensure the trend hasn’t changed. In our unpredictable world, trends might shift – weekends could start on Monday, a four-day workweek might become the norm, or perhaps Elon Musk will have us working seven days a week. Whatever happens, with the right data-driven approach and regular updates, we’ll always be prepared to adapt efficiently!

*ofer, if you’ve read this far, thank you so much for your patience, and I wish you a wonderful day!* 😊